From f955453f8e5adacf92f53966f8f0f86376b5a122 Mon Sep 17 00:00:00 2001
From: Heiko Schocher <hs@denx.de>
Date: Sat, 1 Feb 2025 09:40:10 +0100
Subject: [PATCH 19/20] shell.py: fix crash when no ParamikoConnector
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

when instance has no ParamikoConnector, copy crashes with

│     File "/home/hs/data/Entwicklung/zug/source/u-boot-test/tbottesting/tbottest/../tbot/tbot/machine/linux/copy.py", line 163, in copy
│       isinstance(p2.host, connector.ParamikoConnector)
│   AttributeError: module 'tbot.machine.connector' has no attribute 'ParamikoConnector'

Fix this.

Signed-off-by: Heiko Schocher <hs@denx.de>
---
 tbot/machine/linux/copy.py | 31 +++++++++++++++++++++----------
 1 file changed, 21 insertions(+), 10 deletions(-)

diff --git a/tbot/machine/linux/copy.py b/tbot/machine/linux/copy.py
index 9892e0d..51a189c 100644
--- a/tbot/machine/linux/copy.py
+++ b/tbot/machine/linux/copy.py
@@ -126,12 +126,29 @@ def copy(p1: linux.Path[H1], p2: linux.Path[H2], remote_copy=False) -> None:
 
     .. versionadded:: 0.10.2
     """
+    try:
+        p1isParamiko = isinstance(p1.host, connector.ParamikoConnector)
+    except:
+        p1isParamiko = False
+    try:
+        p1isSSH = isinstance(p1.host, connector.SSHConnector)
+    except:
+        p1isSSH = False
+    try:
+        p2isParamiko = isinstance(p2.host, connector.ParamikoConnector)
+    except:
+        p2isParamiko = False
+    try:
+        p2isSSH = isinstance(p2.host, connector.SSHConnector)
+    except:
+        p2isSSH = False
+
     if isinstance(p1.host, p2.host.__class__) or isinstance(p2.host, p1.host.__class__):
         # Both paths are on the same host
         p2_w1 = linux.Path(p1.host, p2)
         p1.host.exec0("cp", p1, p2_w1)
         return
-    elif isinstance(p1.host, connector.SSHConnector) and p1.host.host is p2.host and (remote_copy == False):
+    elif p1isSSH and p1.host.host is p2.host and (remote_copy == False):
         # Copy from an SSH machine
         _scp_copy(
             local_path=p2,
@@ -145,7 +162,7 @@ def copy(p1: linux.Path[H1], p2: linux.Path[H2], remote_copy=False) -> None:
             authenticator=p1.host.authenticator,
             use_multiplexing=p1.host.use_multiplexing,
         )
-    elif isinstance(p2.host, connector.SSHConnector) and p2.host.host is p1.host:
+    elif p2isSSH and p2.host.host is p1.host:
         # Copy to an SSH machine
         _scp_copy(
             local_path=p1,
@@ -159,10 +176,7 @@ def copy(p1: linux.Path[H1], p2: linux.Path[H2], remote_copy=False) -> None:
             authenticator=p2.host.authenticator,
             use_multiplexing=p2.host.use_multiplexing,
         )
-    elif isinstance(p1.host, connector.SubprocessConnector) and (
-        isinstance(p2.host, connector.ParamikoConnector)
-        or isinstance(p2.host, connector.SSHConnector)
-    ):
+    elif isinstance(p1.host, connector.SubprocessConnector) and (p2isParamiko or p2isSSH):
         # Copy from local to ssh labhost
         _scp_copy(
             local_path=p1,
@@ -176,10 +190,7 @@ def copy(p1: linux.Path[H1], p2: linux.Path[H2], remote_copy=False) -> None:
             authenticator=p2.host.authenticator,
             use_multiplexing=p2.host.use_multiplexing,
         )
-    elif isinstance(p2.host, connector.SubprocessConnector) and (
-        isinstance(p1.host, connector.ParamikoConnector)
-        or isinstance(p1.host, connector.SSHConnector)
-    ):
+    elif isinstance(p2.host, connector.SubprocessConnector) and (p1isParamiko or p1isSSH):
         # Copy to local from ssh labhost
         _scp_copy(
             local_path=p2,
-- 
2.31.1

